models.py

```diff
@@ -62,6 +63,86 @@ def transform(self, datagen, validation_datagen=None):
+        self.model = DCAN(**architecture_config['model_params'])
+        self.weight_regularization = weight_regularization_unet
+        self.optimizer = optim.Adam(self.weight_regularization(self.model, **architecture_config['regularizer_params']),
+                                    **architecture_config['optimizer_params'])
+        self.loss_function = [('mask', segmentation_loss, architecture_config['loss_weights']['mask']),
+                              ('contour', segmentation_loss, architecture_config['loss_weights']['contour']),
+                              ('mask_auxiliary_classifiers', list_segmentation_loss,
+                               architecture_config['loss_weights']['mask_auxiliary_classifiers']),
+                              ('contour_auxiliary_classifiers', list_segmentation_loss,
+                               architecture_config['loss_weights']['contour_auxiliary_classifiers'])]
+        self.callbacks = callbacks_dcan(self.callbacks_config)
+
+    def transform(self, datagen, validation_datagen=None):
+        outputs = self._transform(datagen, validation_datagen)
+        for name, prediction in outputs.items():
+            if "auxiliary" in name:
+                continue
+            prediction_ = [sigmoid(np.squeeze(mask)) for mask in prediction]
+            outputs[name] = np.array(prediction_)
+        return outputs
+
+    def _transform(self, datagen, validation_datagen=None):
+        self.model.eval()
+        batch_gen, steps = datagen
+        outputs = {}
+        for batch_id, data in enumerate(batch_gen):
+            if isinstance(data, list):
+                X = data[0]
+            else:
+                X = data
+
+            if torch.cuda.is_available():
+                X = torch.autograd.Variable(X, volatile=True).cuda()
+            else:
+                X = torch.autograd.Variable(X, volatile=True)
+
+            outputs_batch = self.model(X)
+            for name, output in zip(self.output_names, outputs_batch):
+                if "auxiliary" in name:
+                    continue
+                output_ = output.data.cpu().numpy()
+                outputs.setdefault(name, []).append(output_)
+            if batch_id == steps:
+                break
+        outputs = {'{}_prediction'.format(name): np.vstack(outputs_) for name, outputs_ in outputs.items() if "auxiliary" not in name}
+        return outputs
+
+    def _fit_loop(self, data):
+        X = data[0]
+        targets_tensors = data[1:]
+
+        if torch.cuda.is_available():
+            X = torch.autograd.Variable(X).cuda()
+            targets_var = []
+            for target_tensor in targets_tensors:
+                targets_var.append(torch.autograd.Variable(target_tensor).cuda())
+        else:
+            X = torch.autograd.Variable(X)
+            targets_var = []
+            for target_tensor in targets_tensors:
+                targets_var.append(torch.autograd.Variable(target_tensor))
```

Comment:
@apyskir you are repeating yourself here. if you do if torch.cuda the X = this or that first then you can have this for loop written just once

Reply:
Yes, just I saw things like this: https://github.com/neptune-ml/open-solution-data-science-bowl-2018/blob/final-solution/steps/pytorch/models.py#L52 and I just kind of copied the pattern. I agree that it is not the best possible form of writing it.